Sources/PackageCollections/PackageIndex.swift

```diff
@@ -0,0 +1,250 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import Foundation
+import PackageModel
+import TSCBasic
+
+struct PackageIndex: PackageIndexProtocol, Closable {
+    private let configuration: PackageIndexConfiguration
+    private let httpClient: HTTPClient
+    private let callbackQueue: DispatchQueue
+    private let observabilityScope: ObservabilityScope
+    
+    private let decoder: JSONDecoder
+
+    private let cache: SQLiteBackedCache<CacheValue>?
+    
+    var isEnabled: Bool {
+        self.configuration.enabled && self.configuration.url != .none
+    }
+
+    init(
+        configuration: PackageIndexConfiguration,
+        customHTTPClient: HTTPClient? = nil,
+        callbackQueue: DispatchQueue,
+        observabilityScope: ObservabilityScope
+    ) {
+        self.configuration = configuration
+        self.httpClient = customHTTPClient ?? HTTPClient()
+        self.callbackQueue = callbackQueue
+        self.observabilityScope = observabilityScope
+        
+        self.decoder = JSONDecoder.makeWithDefaults()
+        
+        if configuration.cacheTTLInSeconds > 0 {
+            var cacheConfig = SQLiteBackedCacheConfiguration()
+            cacheConfig.maxSizeInMegabytes = configuration.cacheSizeInMegabytes
+            self.cache = SQLiteBackedCache<CacheValue>(
+                tableName: "package_index_cache",
+                path: configuration.cacheDirectory.appending(component: "index-package-metadata.db"),
+                configuration: cacheConfig
+            )
+        } else {
+            self.cache = nil
+        }
+    }
+    
+    func close() throws {
+        try self.cache?.close()
+    }
+
+    func getPackageMetadata(
+        identity: PackageIdentity,
+        location: String?,
+        callback: @escaping (Result<PackageCollectionsModel.PackageMetadata, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { url, callback in
+            if let cached = try? self.cache?.get(key: identity.description),
+               cached.dispatchTime + DispatchTimeInterval.seconds(self.configuration.cacheTTLInSeconds) > DispatchTime.now() {
+                return callback(.success((package: cached.package, collections: [], provider: self.createContext(host: url.host, error: nil))))
+            }
+            
+            // TODO: rdar://87582621 call package index's get metadata API
+            let metadataURL = url.appendingPathComponent("packages").appendingPathComponent(identity.description)
+            self.httpClient.get(metadataURL) { result in
```

Comment:
for later PRs: we would probably want to set a retry / exponential backoff strategy